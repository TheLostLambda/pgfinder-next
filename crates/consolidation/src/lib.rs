use std::io::Cursor;

use polars::prelude::*;

// Constants ===========================================================================================================

// NOTE: Z-value for 95% confidence interval
const Z_SCORE_FOR_CI_95: f64 = 1.96;
const OLIGO_STATE_REGEX: &str = r"\|(\d+)";

struct InputColumns;
impl InputColumns {
    const ALL: [&str; 5] = [
        Self::STRUCTURE,
        Self::INTENSITY,
        Self::RT,
        Self::THEO,
        Self::PPM,
    ];
    const STRUCTURE: &str = "Structure";
    const INTENSITY: &str = "Consolidated Intensity";
    const RT: &str = "Consolidated RT (min)";
    const THEO: &str = "Consolidated Theo (Da)";
    const PPM: &str = "Consolidated Delta (ppm)";
}

struct TemporaryColumns;
impl TemporaryColumns {
    const REPLICATE: &str = "Replicate";
    const AVG_INTENSITY: &str = "Avg Intensity";
    const CI_INTENSITY: &str = "95% CI Intensity";
}

struct OutputColumns;
impl OutputColumns {
    const STRUCTURE: &str = "Structure";
    const OLIGO_STATE: &str = "Oligo State";
    const AVG_ABUNDANCE: &str = "Average Abundance (%)";
    const CI_ABUNDANCE: &str = "95% CI Abundance (%)";
    const AVG_RT: &str = "Average RT (min)";
    const CI_RT: &str = "95% CI RT (min)";
    const THEO: &str = "Theo (Da)";
    const PPM: &str = "Absolute Delta (ppm)";
    const PRESENT_IN: &str = "Present In";
    const REPLICATE_INTENSITY: &str = "Intensity";
    const REPLICATE_ABUNDANCE: &str = "Abundance";
}

// Public API ==========================================================================================================

pub type ReplicateNumber = u32;
pub type Replicate = (ReplicateNumber, String);

pub fn consolidate_replicates(replicates: &[Replicate]) -> String {
    // SAFETY: Overall this unwrap should be okay, since I also control PGFinder's outputs
    load_replicates(replicates)
        .map(consolidate)
        .and_then(into_csv)
        .unwrap_or_default()
}

// Private Types =======================================================================================================

struct ReplicateFrame(Vec<ReplicateNumber>, LazyFrame);

// Private Functions ===================================================================================================

// TODO: Use a fixed-point Decimal type instead of a float, so that we don't have nasty floating-point errors that show
// up in all of the results...
fn load_replicates(replicates: &[Replicate]) -> PolarsResult<ReplicateFrame> {
    let replicate_numbers: Vec<_> = replicates.iter().map(|&(n, _)| n).collect();
    let lazyframe = concat(
        replicates
            .iter()
            .map(|&(replicate, ref csv)| {
                Ok(CsvReader::new(Cursor::new(csv))
                    .finish()?
                    .lazy()
                    .select(InputColumns::ALL.map(col))
                    .drop_nulls(None)
                    .with_column(lit(replicate).alias(TemporaryColumns::REPLICATE)))
            })
            .collect::<PolarsResult<Vec<_>>>()?,
        UnionArgs::default(),
    );
    Ok(ReplicateFrame(replicate_numbers, lazyframe?))
}

// FIXME: Benchmark if taking in the replicate numbers is even worth it... I'm just trying to not break the chain of
// lazy operations here, so I'm using `group_by` instead of just `pivot`...
fn consolidate(ReplicateFrame(replicates, df): ReplicateFrame) -> LazyFrame {
    let individual_replicate_columns = df
        .clone()
        .group_by([InputColumns::STRUCTURE])
        .agg(wide_formatted_replicate_columns(&replicates))
        .with_columns(replicate_abundance_columns(&replicates));

    df.group_by([InputColumns::STRUCTURE])
        .agg(consolidated_columns())
        .with_column(oligo_state_column(OLIGO_STATE_REGEX))
        .with_columns(average_abundance_columns())
        .join(
            individual_replicate_columns,
            [col(OutputColumns::STRUCTURE)],
            [col(InputColumns::STRUCTURE)],
            JoinArgs::default(),
        )
        .select(formatted_output_columns(&replicates))
        .sort(
            [
                OutputColumns::OLIGO_STATE,
                OutputColumns::AVG_ABUNDANCE,
                OutputColumns::THEO,
            ],
            SortMultipleOptions::new().with_order_descending_multi([false, true, false]),
        )
}

fn into_csv(df: LazyFrame) -> PolarsResult<String> {
    let mut result = Vec::new();
    CsvWriter::new(&mut result).finish(&mut df.collect()?)?;
    // SAFETY: The `CsvWriter` should always return valid UTF-8
    Ok(String::from_utf8(result).unwrap())
}

// ---------------------------------------------------------------------------------------------------------------------

fn wide_formatted_replicate_columns(replicates: &[u32]) -> Vec<Expr> {
    [
        vec![len().alias(OutputColumns::PRESENT_IN)],
        replicates
            .iter()
            .map(|&n| {
                col(InputColumns::INTENSITY)
                    .filter(col(TemporaryColumns::REPLICATE).eq(n))
                    .first()
                    .alias(replicate_column(n, OutputColumns::REPLICATE_INTENSITY))
            })
            .collect(),
    ]
    .concat()
}

fn replicate_abundance_columns(replicates: &[u32]) -> Vec<Expr> {
    replicates
        .iter()
        .map(|&n| {
            let replicate_intensity_column =
                replicate_column(n, OutputColumns::REPLICATE_INTENSITY);
            intensity_to_abundance(&replicate_intensity_column, &replicate_intensity_column)
                .alias(replicate_column(n, OutputColumns::REPLICATE_ABUNDANCE))
        })
        .collect()
}

fn consolidated_columns() -> [Expr; 6] {
    let confidence_interval = |c| lit(Z_SCORE_FOR_CI_95) * col(c).std(1) / len().sqrt();
    [
        mean(InputColumns::INTENSITY).alias(TemporaryColumns::AVG_INTENSITY),
        confidence_interval(InputColumns::INTENSITY).alias(TemporaryColumns::CI_INTENSITY),
        mean(InputColumns::RT).alias(OutputColumns::AVG_RT),
        confidence_interval(InputColumns::RT).alias(OutputColumns::CI_RT),
        col(InputColumns::THEO).first().alias(OutputColumns::THEO),
        col(InputColumns::PPM)
            .abs()
            .mean()
            .alias(OutputColumns::PPM),
    ]
}

fn oligo_state_column(oligo_state_re: &str) -> Expr {
    col(OutputColumns::STRUCTURE)
        .str()
        .extract(lit(oligo_state_re), 1)
        .alias(OutputColumns::OLIGO_STATE)
}

fn average_abundance_columns() -> [Expr; 2] {
    [
        intensity_to_abundance(
            TemporaryColumns::AVG_INTENSITY,
            TemporaryColumns::AVG_INTENSITY,
        )
        .alias(OutputColumns::AVG_ABUNDANCE),
        intensity_to_abundance(
            TemporaryColumns::CI_INTENSITY,
            TemporaryColumns::AVG_INTENSITY,
        )
        .alias(OutputColumns::CI_ABUNDANCE),
    ]
}

fn formatted_output_columns(replicates: &[u32]) -> Vec<Expr> {
    [
        vec![
            col(OutputColumns::STRUCTURE)
                .str()
                .replace_all(lit(OLIGO_STATE_REGEX), lit(""), false),
            col(OutputColumns::OLIGO_STATE),
            col(OutputColumns::AVG_ABUNDANCE).round(3),
            col(OutputColumns::CI_ABUNDANCE).round(3),
            col(OutputColumns::AVG_RT).round(2),
            col(OutputColumns::CI_RT).round(2),
            col(OutputColumns::THEO).round(4),
            col(OutputColumns::PPM).round(1),
            col(OutputColumns::PRESENT_IN),
        ],
        replicates
            .iter()
            .flat_map(|&n| {
                [
                    col(replicate_column(n, OutputColumns::REPLICATE_INTENSITY)),
                    col(replicate_column(n, OutputColumns::REPLICATE_ABUNDANCE)).round(3),
                ]
            })
            .collect(),
    ]
    .concat()
}

// ---------------------------------------------------------------------------------------------------------------------

fn replicate_column(replicate: u32, column: &str) -> String {
    format!("Replicate {replicate} ({column})")
}

fn intensity_to_abundance(column: &str, total_column: &str) -> Expr {
    lit(100) * col(column) / sum(total_column)
}

// Unit Tests ==========================================================================================================

#[cfg(test)]
mod tests {
    use std::sync::LazyLock;

    use insta::assert_snapshot;

    use super::*;

    static REPLICATES: LazyLock<[Replicate; 3]> = LazyLock::new(|| {
        [
            (1, include_str!("../tests/data/E. coli WT_01.csv")),
            (2, include_str!("../tests/data/E. coli WT_02.csv")),
            (3, include_str!("../tests/data/E. coli WT_03.csv")),
        ]
        .map(|(n, c)| (n, c.to_owned()))
    });

    #[test]
    fn test_load_replicates() {
        assert_snapshot!(
            load_replicates(REPLICATES.as_ref())
                .and_then(|ReplicateFrame(_, df)| into_csv(df))
                .unwrap()
        );
    }

    #[test]
    fn test_consolidate() {
        assert_snapshot!(
            load_replicates(REPLICATES.as_ref())
                .map(consolidate)
                .and_then(into_csv)
                .unwrap()
        );
    }

    #[test]
    fn test_consolidate_replicates() {
        assert_snapshot!(consolidate_replicates(REPLICATES.as_ref()));
    }
}
