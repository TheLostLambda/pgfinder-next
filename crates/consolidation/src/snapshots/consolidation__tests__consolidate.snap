---
source: crates/consolidation/src/lib.rs
expression: load_replicates(&REPLICATES).map(consolidate).and_then(into_csv).unwrap()
---
Structure,Oligo State,Average Abundance (%),95% CI Abundance (%),Average RT (min),95% CI RT (min),Theo (Da),Absolute Delta (ppm),Present In,Replicate 1 (Intensity),Replicate 1 (Abundance),Replicate 2 (Intensity),Replicate 2 (Abundance),Replicate 3 (Intensity),Replicate 3 (Abundance)
gm,0,4.825,2.313,3.62,0.01,498.2061,2.5,3,31944113.0,5.834,14158755.0,3.346,19002675.0,5.063
gm-gm,0,0.617,0.643,10.11,0.04,976.386,2.2,3,911535.0,0.166,5689163.0,1.345,1723056.0,0.459
gm (Anh),0,0.343,0.117,8.21,2.65,478.1799,2.9,3,2077721.0,0.379,1225341.0,0.29,1320874.0,0.352
gm-gm (Anh),0,0.044,,6.8,,956.3598,5.3,1,198699.0,0.036,,,,
gm-AEJA,1,35.075,5.373,10.04,0.06,941.4077,2.8,3,180156373.0,32.904,155495246.0,36.749,137622224.0,36.671
gm-AEJ,1,13.148,3.598,6.57,0.02,870.3706,2.9,3,74594397.0,13.624,56436454.0,13.338,46379829.0,12.358
gm-AEJK,1,5.632,1.306,8.36,0.08,998.4656,2.7,3,31297651.0,5.716,22852779.0,5.401,21844338.0,5.821
gm-AE,1,2.607,1.116,9.57,0.06,698.2858,3.1,3,16500510.0,3.014,10932398.0,2.584,7737109.0,2.062
gm-AEJ (Anh),1,0.963,0.297,13.99,0.02,850.3444,2.2,3,5358618.0,0.979,4598678.0,1.087,3039575.0,0.81
gm-AEJA (Anh),1,0.825,0.271,16.55,0.01,921.3815,2.0,3,4368964.0,0.798,4296720.0,1.015,2466850.0,0.657
gm-AEJG,1,0.65,0.233,7.86,0.08,927.3921,2.4,3,3972772.0,0.726,2570038.0,0.607,2223717.0,0.593
gm-A,1,0.625,0.176,8.03,0.06,569.2432,2.4,3,2430860.0,0.444,2383796.0,0.563,3619062.0,0.964
"gm-AEJQ,   gm-AEJAG",1,0.217,0.077,9.45,0.07,998.4292,3.1,3,1322657.0,0.242,880794.0,0.208,730417.0,0.195
gm-AEJAK,1,0.137,0.015,9.48,0.08,1069.5027,3.3,3,681181.0,0.124,566144.0,0.134,601392.0,0.16
gm-AEJK (Anh),1,0.131,0.029,14.4,0.01,978.4394,2.6,3,670227.0,0.122,633717.0,0.15,457639.0,0.122
gm-AE (Anh),1,0.1,0.043,17.44,0.02,678.2596,2.8,3,601632.0,0.11,482915.0,0.114,261507.0,0.07
gm-AEJN,1,0.097,0.026,5.91,0.02,984.4135,2.6,3,544988.0,0.1,420217.0,0.099,337171.0,0.09
gm-AEJF,1,0.089,0.034,20.96,0.01,1017.439,2.2,3,533134.0,0.097,403697.0,0.095,265131.0,0.071
gm-AEJAA,1,0.065,0.033,11.38,0.06,1012.4448,2.3,3,443919.0,0.081,230707.0,0.055,208427.0,0.056
gm-AEJAN,1,0.062,0.019,8.28,0.08,1055.4506,2.5,3,367394.0,0.067,225329.0,0.053,247256.0,0.066
gm-AEJT,1,0.054,0.014,8.47,0.08,971.4183,2.5,3,307779.0,0.056,218104.0,0.052,201139.0,0.054
gm-AEJAT,1,0.045,0.032,6.02,2.29,1042.4554,2.2,3,349023.0,0.064,153278.0,0.036,108624.0,0.029
gm-AEJS,1,0.044,0.016,6.79,0.02,957.4026,3.3,3,273284.0,0.05,171396.0,0.041,151459.0,0.04
gm-A (Anh),1,0.038,0.006,16.72,0.01,549.217,2.4,3,162889.0,0.03,157011.0,0.037,198924.0,0.053
gm-AEJH,1,0.025,0.01,5.69,0.02,1007.4295,2.6,3,118358.0,0.022,72870.0,0.017,148530.0,0.04
gm-AEJAH,1,0.022,0.005,7.96,0.1,1078.4666,2.4,2,109228.0,0.02,85717.0,0.02,,
gm-AEJV,1,0.008,,15.12,,969.439,3.0,1,,,37116.0,0.009,,
"gm-AEJAA=gm-AEJ,   gm-AEJA=gm-AEJA",2,20.024,3.528,16.01,0.02,1864.8046,2.3,3,105836569.0,19.33,85352267.0,20.172,79001487.0,21.051
"gm-AEJA=gm-AEJ,   gm-AEJ=gm-AEJA",2,3.478,1.4,15.08,0.03,1793.7675,2.6,3,22020529.0,4.022,13139810.0,3.105,11770579.0,3.136
"gm-AEJK=gm-AEJA,   gm-AEJAK=gm-AEJ",2,1.34,0.341,14.49,0.01,1921.8625,2.7,3,7584787.0,1.385,5374634.0,1.27,5116528.0,1.363
"gm-AEJAA=gm-AEJ (Anh),   gm-AEJA=gm-AEJA (Anh)",2,1.258,0.368,20.56,0.01,1844.7784,2.4,3,7202583.0,1.316,5472178.0,1.293,4292906.0,1.144
"gm-AEJ=gm-AEJ,   gm-AEJAT=gm-AE",2,0.664,0.264,14.23,0.01,1722.7304,3.0,3,4164790.0,0.761,2649984.0,0.626,2151119.0,0.573
"gm-AEJA=gm-AEJ (Anh),   gm-AEJ=gm-AEJA (Anh)",2,0.55,0.157,19.66,0.02,1773.7413,2.1,3,3191122.0,0.583,2083631.0,0.492,2140978.0,0.57
"gm-AEJQ=gm-AEJ,   gm-AEJG=gm-AEJA,   gm-AEJAG=gm-AEJ",2,0.479,0.161,14.67,0.02,1850.789,2.4,3,2888709.0,0.528,1873882.0,0.443,1701092.0,0.453
gm-AEJS=gm-AEJA (Anh),2,0.221,0.132,18.7,0.47,1860.7733,1.4,2,691174.0,0.126,1296040.0,0.306,,
"gm-AEJ=gm-AEJ (Anh),   gm-AEJAT=gm-AE (Anh)",2,0.203,0.082,18.77,0.02,1702.7042,4.5,3,1271074.0,0.232,840801.0,0.199,633540.0,0.169
"gm-AEJQ=gm-AEJA,   gm-AEJAG=gm-AEJA",2,0.203,0.052,15.33,0.02,1921.8261,2.2,3,1129627.0,0.206,890576.0,0.21,720451.0,0.192
"gm-AEJK=gm-AEJA (Anh),   gm-AEJAK=gm-AEJ (Anh)",2,0.173,0.053,18.68,0.05,1901.8363,2.1,3,759058.0,0.139,574312.0,0.136,996905.0,0.266
gm-AEJK=gm-AEJ,2,0.172,0.062,13.8,0.01,1850.8254,2.4,3,1049929.0,0.192,686284.0,0.162,578244.0,0.154
gm-AEJAA=gm-AEJA,2,0.109,0.032,16.54,0.02,1935.8417,2.1,3,631527.0,0.115,440573.0,0.104,392650.0,0.105
gm-AEJAK=gm-AEJA,2,0.106,0.031,14.05,0.01,1992.8996,2.2,3,615586.0,0.112,435780.0,0.103,384364.0,0.102
gm-AEJS=gm-AEJA,2,0.092,0.022,14.62,0.02,1880.7995,2.6,3,512615.0,0.094,383081.0,0.091,350121.0,0.093
gm-AEJAN=gm-AEJA,2,0.088,0.027,14.65,0.02,1978.8475,2.8,3,514554.0,0.094,358341.0,0.085,312938.0,0.083
gm-AEJS=gm-AEJ (Anh),2,0.072,,18.08,,1789.7362,2.0,1,324881.0,0.059,,,,
gm-AEJG=gm-AEJ,2,0.064,0.022,13.91,0.01,1779.7519,2.7,3,385624.0,0.07,261702.0,0.062,216452.0,0.058
"gm-AEJN=gm-AEJA,   gm-AEJAN=gm-AEJ",2,0.06,0.017,13.89,0.02,1907.8104,2.5,3,345605.0,0.063,244727.0,0.058,214016.0,0.057
gm-AEJK=gm-AEJ (Anh),2,0.051,0.01,17.94,0.15,1830.7992,1.4,3,273935.0,0.05,207632.0,0.049,200763.0,0.053
"gm-AEJQ=gm-AEJ (Anh),   gm-AEJG=gm-AEJA (Anh),   gm-AEJAG=gm-AEJ (Anh)",2,0.03,0.01,19.12,0.01,1830.7628,0.8,3,175800.0,0.032,126144.0,0.03,98832.0,0.026
"gm-AEJQ=gm-AEJA (Anh),   gm-AEJAG=gm-AEJA (Anh)",2,0.028,0.01,19.73,0.02,1901.7999,2.1,3,168161.0,0.031,118403.0,0.028,85583.0,0.023
gm-AEJF=gm-AEJA,2,0.025,0.009,24.38,0.0,1940.8359,1.4,3,151324.0,0.028,97109.0,0.023,83854.0,0.022
gm-AEJAA=gm-AEJA (Anh),2,0.022,0.004,21.17,0.03,1915.8155,1.8,3,113937.0,0.021,86431.0,0.02,95253.0,0.025
gm-AEJS=gm-AEJ,2,0.015,,13.55,,1809.7624,0.3,1,67851.0,0.012,,,,
"gm-AEJH=gm-AEJA,   gm-AEJAH=gm-AEJ",2,0.013,0.005,13.41,0.01,1930.8264,2.5,2,69066.0,0.013,44315.0,0.01,,
gm-AEJAK=gm-AEJA (Anh),2,0.01,,18.16,,1972.8734,4.2,1,43008.0,0.008,,,,
"gm-AEJN=gm-AEJA (Anh),   gm-AEJAN=gm-AEJ (Anh)",2,0.009,,18.26,,1887.7842,1.2,1,41116.0,0.008,,,,
gm-AEJF=gm-AEJ,2,0.008,,23.79,,1869.7988,1.4,1,34087.0,0.006,,,,
gm-AEJH=gm-AEJA,2,0.008,,13.43,,1930.8264,2.0,1,,,,,34702.0,0.009
gm-AEJAT=gm-AEJA,2,0.008,0.0,15.87,0.03,1965.8523,1.9,2,33480.0,0.006,,,34903.0,0.009
gm-AEJN=gm-AEJ,2,0.007,0.002,13.23,0.02,1836.7733,2.8,2,38643.0,0.007,28073.0,0.007,,
"gm-AEJAA=gm-AEJ=gm-AEJA,   gm-AEJA=gm-AEJA=gm-AEJA",3,2.496,1.108,18.86,0.01,2788.2015,3.5,3,16268983.0,2.971,9277698.0,2.193,8132646.0,2.167
"gm-AEJAA=gm-AEJ=gm-AEJ,   gm-AEJA=gm-AEJ=gm-AEJA,   gm-AEJ=gm-AEJA=gm-AEJA",3,0.531,0.196,18.23,0.29,2717.1644,2.2,3,3253771.0,0.594,2169690.0,0.513,1741631.0,0.464
"gm-AEJAA=gm-AEJ=gm-AEJA (Anh),   gm-AEJA=gm-AEJA=gm-AEJA (Anh)",3,0.319,0.135,22.39,0.02,2768.1753,2.3,3,2054169.0,0.375,1132610.0,0.268,1121176.0,0.299
"gm-AEJAA=gm-AEJ=gm-AEJ (Anh),   gm-AEJA=gm-AEJ=gm-AEJA (Anh),   gm-AEJ=gm-AEJA=gm-AEJA (Anh)",3,0.168,0.049,21.6,0.03,2697.1382,1.8,3,977802.0,0.179,685660.0,0.162,604518.0,0.161
"gm-AEJK=gm-AEJA=gm-AEJA,   gm-AEJAK=gm-AEJ=gm-AEJA",3,0.093,0.033,17.41,0.02,2845.2594,2.2,3,565615.0,0.103,333994.0,0.079,349719.0,0.093
"gm-AEJQ=gm-AEJ=gm-AEJA,   gm-AEJG=gm-AEJA=gm-AEJA,   gm-AEJAG=gm-AEJ=gm-AEJA",3,0.056,0.022,17.72,0.02,2774.1859,1.4,3,350259.0,0.064,208093.0,0.049,190947.0,0.051
"gm-AEJK=gm-AEJA=gm-AEJA (Anh),   gm-AEJAK=gm-AEJ=gm-AEJA (Anh)",3,0.049,0.013,20.76,0.05,2825.2332,1.7,3,226903.0,0.041,266015.0,0.063,167093.0,0.045
"gm-AEJA=gm-AEJ=gm-AEJ,   gm-AEJ=gm-AEJ=gm-AEJA",3,0.042,0.021,17.45,0.01,2646.1273,1.7,3,281942.0,0.051,155367.0,0.037,127311.0,0.034
"gm-AEJA=gm-AEJ=gm-AEJ (Anh),   gm-AEJ=gm-AEJ=gm-AEJA (Anh)",3,0.036,0.009,21.16,0.02,2626.1011,1.9,3,201138.0,0.037,162098.0,0.038,128863.0,0.034
"gm-AEJQ=gm-AEJA=gm-AEJA,   gm-AEJAG=gm-AEJA=gm-AEJA",3,0.031,0.01,18.25,0.02,2845.223,2.0,3,181581.0,0.033,128442.0,0.03,107984.0,0.029
gm-AEJAA=gm-AEJA=gm-AEJA,3,0.021,0.011,19.23,0.02,2859.2386,1.3,2,119064.0,0.022,66403.0,0.016,,
"gm-AEJQ=gm-AEJ=gm-AEJ,   gm-AEJG=gm-AEJ=gm-AEJA,   gm-AEJAG=gm-AEJ=gm-AEJ",3,0.019,,17.18,,2703.1488,2.0,1,85854.0,0.016,,,,
"gm-AEJQ=gm-AEJ=gm-AEJA (Anh),   gm-AEJG=gm-AEJA=gm-AEJA (Anh),   gm-AEJAG=gm-AEJ=gm-AEJA (Anh)",3,0.016,0.015,21.21,0.03,2754.1597,1.2,2,104983.0,0.019,36379.0,0.009,,
gm-AEJAN=gm-AEJA=gm-AEJA,3,0.015,,17.65,,2902.2444,2.2,1,68147.0,0.012,,,,
"gm-AEJK=gm-AEJ=gm-AEJA,   gm-AEJAK=gm-AEJ=gm-AEJ",3,0.013,0.004,16.94,0.02,2774.2223,1.7,3,71676.0,0.013,57027.0,0.013,42901.0,0.011
"gm-AEJQ=gm-AEJA=gm-AEJA (Anh),   gm-AEJAG=gm-AEJA=gm-AEJA (Anh)",3,0.012,0.008,21.77,0.02,2825.1968,2.8,2,74611.0,0.014,,,36716.0,0.01
gm-AEJS=gm-AEJA=gm-AEJA,3,0.008,,17.83,,2804.1964,1.7,1,,,,,36440.0,0.01
